### PR TITLE
Order payment method title matches payment method used instead of gateway title

### LINKED ISSUE #2057: Order complete page should show correct payment method
On the order complete page, currently with UPE the payment method shown is the name of the gateway, not the specific payment method used in the transaction. 

Order complete screen shows inaccurate payment method
_Order complete screen shows payment gateway title instead of payment method used_

[This section of code](https://github.com/woocommerce/woocommerce/blob/trunk/templates/checkout/order-receipt.php#L39) on the order complete page calls the function `$order->get_payment_method_title()`. We need to [set this value](https://woocommerce.github.io/code-reference/classes/WC-Order.html#method_set_payment_method_title) correctly to the relevant payment method used in the transaction when we process the payment for the order.

### PR DESCRIPTION
**Pull request by FangedParakeet**
Fixes #2057 

### Update order payment method title to match payment method used in UPE transaction

On the order complete page, currently with UPE the payment method shown is the name of the gateway, not the specific payment method used in the transaction. 

Order complete screen shows inaccurate payment method
_Order complete screen shows payment gateway title instead of payment method used_

This PR updates the code, so that the payment method details from the charge accompanying the associated payment intention are used to accurately display the payment method used for a completed order. Just FYI, I tested the gateway with a 100% discount code and in this case, the entire payment gateway disappears and there is no `PAYMENT METHOD` displayed at all at checkout. Considering this, I am proceeding under the assumption that every order that is processed through the UPE payment gateway _will_ have an accompanying charge of some sort.

#### Testing instructions

1. Please pull the changes in this PR and the most up to date version of the Payments Server.
2. Enable the UPE gateway at checkout by setting the UPE feature flag by adding a row to your `wc_options` table for your site with `option_name` of "_wcpay_feature_upe" and `option_value` of "1". Additionally, if you would like to test this code with the theoretically functional Giropay payment method, please add `giropay` to [this array](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L111).
3. Add an item to your store's cart, proceed to checkout, and complete a checkout using the UPE payment gateway.
4. After successful checkout, on the order complete page, below the order total, you should now see the correct payment method displayed. If you used the Visa test card, the corresponding value should be `Visa credit card (WooCommerce Payments)`; if you used Giropay, you should see `Giropay (WooCommerce Payments)`.
5. That's it. Now use this time to recall a fond yet fleeting memory from your childhood perhaps.

*

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

File: includes/wc-payment-api/models/class-wc-payments-api-intention.php [hunk 1]
```diff
@@ -40,6 +40,13 @@ class WC_Payments_API_Intention {
 	 */
 	private $status;
 
+	/**
+	 * The ID of charge associated with intention
+	 *
+	 * @var string
+	 */
+	private $charge_id;
```
Comment (FangedParakeet) [inline includes/wc-payment-api/models/class-wc-payments-api-intention.php:48]: This doesn't really have anything to do with my changes, but I noticed this private variable was missing...and I couldn't really see any good reason why it shouldn't be there. The code still worked, I guess. But anyways, I thought better to add it.

File: includes/payment-methods/class-upe-payment-gateway.php [hunk 2]
```diff
@@ -285,6 +287,56 @@ public function process_redirect_payment( $order_id, $intent_id, $save_payment_m
...
+					'unknown' => __( 'unknown', 'woocommerce-payments' ),
+				];
+
+				$payment_method_title = sprintf(
+					// Translators: %1$s card brand, %2$s card funding (prepaid, credit, etc.).
+					__( '%1$s %2$s card', 'woocommerce-payments' ),
+					ucfirst( $details['network'] ),
+					$funding_types[ $details['funding'] ]
+				);
+				break;
```
Comment (FangedParakeet) [inline includes/payment-methods/class-upe-payment-gateway.php:318]: This mirrors the formatting from the transaction details page [over here](https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-details/payment-method/card/index.js#L51).

File: includes/payment-methods/class-upe-payment-gateway.php [hunk 3]
```diff
@@ -285,6 +287,56 @@ public function process_redirect_payment( $order_id, $intent_id, $save_payment_m
...
+
+			case 'sofort':
+				$payment_method_title = 'Sofort';
+				break;
+
+			default:
+				return;
+		}
+
+		$order->set_payment_method_title( "$payment_method_title (WooCommerce Payments)" );
```
Comment (FangedParakeet) [inline includes/payment-methods/class-upe-payment-gateway.php:336]: Adding the `(WooCommerce Payments)` suffix follows the format I found [here](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payments-payment-request-button-handler.php#L408), in the only other place I could find the order's payment method title being manually updated, just FYI.

File: includes/payment-methods/class-upe-payment-gateway.php [hunk 4]
```diff
@@ -285,6 +287,56 @@ public function process_redirect_payment( $order_id, $intent_id, $save_payment_m
 		}
 	}
 
+	/**
+	 * Set formatted readable payment method title for order,
+	 * using payment method details from accompanying charge.
+	 *
+	 * @param WC_Order $order WC Order being processed.
+	 * @param array    $payment_method_details Array of payment method details from charge.
```
Comment (FangedParakeet) [inline includes/payment-methods/class-upe-payment-gateway.php:295]: You can find a description of the various shapes this `payment_method_details` can take over [here](https://stripe.com/docs/api/charges/object#charge_object-payment_method_details).

**Review by frosso — APPROVED** [review]
Nice! Changes make sense and it looks like it is backwards compatible, with a fallback in case the payment method type is not recognized 👍 🚢

**Pull request merged**